docs/hugo/content/architecture-decision-records/ADR-2022-01-Reconciler-Extensions.md

```diff
@@ -0,0 +1,106 @@
+
+## Decision
+
+To provide a stable implementation location for extensions, we'll introduce a dedicated package alongside the existing versioned packages for each supported service. This package will provide extension support for all supported resource versions. For example, alongside the existing **batch** packages `v1alpha1api20210101`, and `v1alpha1api20210101storage`, we will generate `extensions`.
+
+Into the `extensions` package we will codegen a skeleton framework. For each resource there will be a separate extension type, suffixed with `Extensions`. For example, for the **compute** resources `Disk`, `VirtualMachine`, and `VirtualMachineScaleSet` we will create the extension types `DiskExtensions`, `VirtualMachineExtensions`, and `VirtualMachineScaleSetExtensions`.
+
+We'll enhance the existing registration file (`controller_resources_gen.go`) to also provide registration of extensions. They will be indexed by the GVK of each resource, allowing easy lookup by the generic reconciler.
+
+For each supported extension, we'll define an interface. Implementers will assert type compatibility with this extension interface to ensure they have the correct method signature. The generic reconciler will identify available extensions by testing for the presence of the interface.
+
+### Example
+
+To allow customization of error handling for Azure Redis, we'll introduce an extension point, which would look something like this:
+
+We create the required extension interface in `genruntime`:
+
+``` go
+type ErrorClassifierExtension interface {
+    // Classify the provided error.
+    // cloudError is the error returned from ARM.
+    // details contains the initial analysis by the ASO reconciler, and may be changed if desired.
+    // apiVersion is the ARM API version used for the request
+    ClassifyError(cloudError *genericarmclient.CloudError, details *CloudErrorDetails, apiVersion string)
+}
+```
+
+In the `cache/extensions` package, the new extension point will be implemented:
+
+``` go
+var _ ErrorClassifierExtension = &RedisExtensions{}
+
+func (e *RedisExtensions) ClassifyError(
+    cloudError *genericarmclient.CloudError, 
+    details *CloudErrorDetails, 
+    apiVersion string) {
+
+    inner := cloudError.InnerError
+    if inner == nil {
+        return
+    }
+
+    code := to.String(inner.Code)
+    if code == "Conflict" {
+        // For Redis, Conflict can be retried as it can occur because something doesn't yet exist
+        details.Classification = CloudErrorRetryable
+    }
+}
+```
+
+This code goes in a different file so that it's not obliterated next time we rerun the generator.
+
+In the generic reconciler, we make use of the extension:
+
+``` go
+func (r *AzureDeploymentReconciler) makeReadyConditionFromError(
+    cloudError *genericarmclient.CloudError) conditions.Condition {
+
+	var severity conditions.ConditionSeverity
+	errorDetails := ClassifyCloudError(cloudError)
+
+    ext := r.FindErrorClassifierExtension()
```

Comment:
I like this not having to deal with "didn't find" or `nil`. Just a no-op method if no extension found makes usage clean.